rfcs/text/0017_url_service.md

```diff
@@ -258,58 +349,80 @@ browser, wherever they choose to use the URL Service:
  * Common URL Service client interface for the server-side and the client-side.
  */
 interface IUrlService {
-  generators: IUrlGeneratorClient;
+  locators: ILocatorClient;
   shortUrls: IShortUrlClient;
 }
 ```
 
-The URL generator client will basically be a registry of URL generators:
+
+### Locators
+
+The locator business logic will be contained in `ILocatorClient` client and will
+provided two main functionalities:
+
+1. It will provide a facility to create locators.
+1. It will also be a registry of locators, every newly created locator is
+   automatically added to the registry.
 
 ```ts
-/**
- * URL generator registry.
- */
-interface IUrlGeneratorClient {
-  register<S>(urlGenerator: UrlGeneratorDefinition<S>): UrlGenerator;
-  get<S>(id: string): UrlGenerator<S>;
+interface ILocatorClient {
+  create<P>(definition: LocatorDefinition<P>): Locator<P>;
+  get<P>(id: string): Locator<P>;
 }
+```
+
+The `LocatorDefinition` interface is a developer-friendly interface for creating
+new locators. Mainly two things will be required from each new locator:
+
+1. Implement the `getLocation()` method, which give the locator specific `params`
+   object returns a Kibana location, see description of `KibanaLocation` below.
+2. Implement the `PersistableState` interface which we use in Kibana. This will
+   allow to migrate the locator `params`. Implementation of the `PersistableState`
+   interface will replace the `.isDeprecated` and `.migrate()` properties of URL
+   generators.
 
-interface UrlGeneratorDefinition<State> {
+
+```ts
+interface LocatorDefinition<P> extends PeristableState<P> {
   id: string;
-  createUrl: (state: State) => Promise<string>;
-  isDeprecated?: boolean;
-  migrate?: (state: State) => Promise<{ state: unknown; id: string }>;
+  getLocation(params: P): KibanaLocation;
 }
+```
 
-interface UrlGenerator<State> {
-  isDeprecated: boolean;
-  createUrl(state: State): Promise<URL>;
-  migrate(): Promise<never | { state: unknown; id: string }>;
+Each constructed locator will have the following interface:
+
+```ts
+interface Locator<P> {
+  /** Creates a new short URL saved object using this locator. */
+  createShortUrl(params: P, slug?: string): Promise<ShortUrl>;
+  /** Returns a relative URL to the client-side redirect endpoint using this locator. */
+  getRedirectPath(params: P): string;
```

Comment:
i would mark this one as internal and not intended to be used by external consumers and mention short urls are the preferred way to deep link from external sources

Reply:
Will do.